client/clientimpl.go

```diff
@@ -0,0 +1,348 @@
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+// Impl is a Client implementation.
+type Impl struct {
+	logger   types.Logger
+	settings StartSettings
+
+	// OpAMP Server URL.
+	url *url.URL
+
+	// HTTP request headers to use when connecting to OpAMP Server.
+	requestHeader http.Header
+
+	// Websocket dialer and connection.
+	dialer    websocket.Dialer
+	conn      *websocket.Conn
+	connMutex sync.RWMutex
+
+	// True is Start() is successful.
+	isStarted bool
+
+	// Context to use for background running go routines.
+	runContext context.Context
+	runCancel  context.CancelFunc
+
+	// True when stopping is in progress.
+	isStoppingFlag  bool
+	isStoppingMutex sync.RWMutex
+
+	// Indicates that the Client is fully stopped.
+	stoppedSignal chan struct{}
+
+	// The sender is responsible for sending portion of the OpAMP protocol.
+	sender *internal.Sender
+}
+
+var _ OpAMPClient = (*Impl)(nil)
+
+const retryAfterHTTPHeader = "Retry-After"
+
+func New(logger types.Logger) *Impl {
+	if logger == nil {
+		logger = &nopLogger{}
+	}
+
+	w := &Impl{
+		logger:        logger,
+		stoppedSignal: make(chan struct{}, 1),
+		sender:        internal.NewSender(logger),
+	}
+	return w
+}
+
+func (w *Impl) Start(settings StartSettings) error {
+	if w.isStarted {
+		return errors.New("already started")
+	}
+
+	w.settings = settings
+
+	var err error
+
+	// Prepare server connection settings.
+	w.url, err = url.Parse(w.settings.OpAMPServerURL)
+	if err != nil {
+		return err
+	}
+
+	w.dialer = *websocket.DefaultDialer
+
+	if w.settings.TLSConfig != nil {
+		w.url.Scheme = "wss"
+	}
+	w.dialer.TLSClientConfig = w.settings.TLSConfig
+
+	if w.settings.AuthorizationHeader != "" {
+		w.requestHeader = http.Header{}
+		w.requestHeader["Authorization"] = []string{w.settings.AuthorizationHeader}
+	}
+
+	// Prepare the first status report.
+	w.sender.UpdateStatus(
+		func(statusReport *protobufs.StatusReport) {
+			statusReport.AgentDescription = &protobufs.AgentDescription{
+				AgentType:    w.settings.AgentType,
+				AgentVersion: w.settings.AgentVersion,
+			}
+
+			statusReport.ServerProvidedAllAddonsHash = w.settings.LastServerProvidedAllAddonsHash
+		},
+	)
+
+	w.startConnectAndRun()
+
+	w.isStarted = true
+
+	return nil
+}
+
+func (w *Impl) Stop(ctx context.Context) error {
+	if !w.isStarted {
+		return errors.New("cannot stop because not started")
+	}
+
+	w.isStoppingMutex.Lock()
+	cancelFunc := w.runCancel
+	w.isStoppingFlag = true
+	w.isStoppingMutex.Unlock()
+
+	cancelFunc()
+
+	// Close connection if any.
+	w.connMutex.RLock()
+	conn := w.conn
+	w.connMutex.RUnlock()
+
+	if conn != nil {
+		conn.Close()
+	}
+	// If we are not connected by this point we may still get connected because
+	// we are racing with tryConnectOnce() func. However, after tryConnectOnce we will
+	// check the isStopping flag in receiverLoop() and will exit receiverLoop, so we will
+	// not deadlock.
+
+	// Wait until stopping is finished.
+	select {
```

Comment:
Are you suggesting to make Stop() non-cancellable, so that it blocks indefinitely long until stopping is done?

Reply:
I am just concerned that we would be leaving work undone which could lead for a bad time.

Reply:
The typical user-case for cancelling Stop is usually the final shutdown of the process. In this case it is likely preferable to make sure Stop() finishes within some reasonable time even if there is work remaining than to risk Stop() blocking the process shutdown infinitely long.
We can carefully review the code to see what can potentially go wrong if Stop() is cancelled (you are right that it may be a concern), but I think it is highly desirable to make Stop() cancellable and not block potentially infinitely.
Let's merge the PR and we can see what needs improving here (I feel we need more test coverage to uncover the edge cases anyway).